src/dvc_render/vega.py

```diff
@@ -137,3 +197,293 @@ def generate_markdown(self, report_path=None) -> str:
                 return f"\n![{self.name}]({src})"
 
         return ""
+
+    def get_revs(self):
+        """
+        Returns all revisions that were collected that have datapoints.
+        """
+        return self.properties.get("revs_with_datapoints", [])
+
+    def _process_optional_anchors(self, split_anchors: List[str]):
+        optional_anchors = [
+            anchor
+            for anchor in [
+                "color",
+                "column",
+                "group_by_x",
+                "group_by_y",
+                "group_by",
+                "pivot_field",
+                "plot_height",
+                "plot_width",
+                "row",
+                "shape",
+                "stroke_dash",
+                "tooltip",
+                "zoom_and_pan",
+            ]
+            if self.template.has_anchor(anchor)
+        ]
+        if not optional_anchors:
+            return None
+
+        self._fill_color(split_anchors, optional_anchors)
+        self._fill_set_encoding(split_anchors, optional_anchors)
```

Comment:
Minor: This is probably outside the scope of this PR and maybe personal bias, but some of why I find the plots code so hard to read is because everything is defined as a class attribute and mutated in place rather than being written in a more functional style. I have no idea what these methods are actually modifying.

Reply:
I can rewrite it as functions and reduce the mutation of class properties if you think that would help.

Reply:
Yes, it's outside the scope of this PR, I just wanted to make a note of it. Let's not expand the scope here.